### PR TITLE
Done rework for va_test.c

### PR DESCRIPTION
**Pull request by SamirMulani**
With older implementations, MAP_CHUNK_SIZE is fixed to 16 GB, as a
result fix value, On low memory the test case is getting failed with
ENOMEM error message where it clearly indicates that not enough memory
is present in the system, which results in a few va_test cases failing.
The above changes will dynamically adjust the MAP_CHUNK_SIZE value
according to the amount of space in the system.

Signed-off-by: Samir Mulani <samir@linux.vnet.ibm.com>

### CODE REVIEW

File: memory/vatest.py.data/va_test.c [hunk 1]
```diff
@@ -324,10 +324,50 @@ void mixed_position_alterhuge(int chunks1, int chunks2, int chunks3)
...
+        int num = 0;
+        FILE* file = popen(cmd, "r");
+        fgets(buffer, 100, file);
+        char * token = strtok(buffer, " ");
+        while( token != NULL )
+        {
+                num = atoi(token);
+                if (num == 0 && token != '0');
+                else{
+                        //printf("Number is: %i", num);
```
Comment (sacsant) [inline memory/vatest.py.data/va_test.c:339]: I guess we don't need this statement and can be removed.

File: memory/vatest.py.data/va_test.c [hunk 2]
```diff
@@ -324,10 +324,50 @@ void mixed_position_alterhuge(int chunks1, int chunks2, int chunks3)
...
+                        token = NULL;
+                        return num;
```
Comment (sacsant) [inline memory/vatest.py.data/va_test.c:341]: A corresponding pclose() call is missing.

File: memory/vatest.py.data/va_test.c [hunk 3]
```diff
@@ -324,10 +324,50 @@ void mixed_position_alterhuge(int chunks1, int chunks2, int chunks3)
...
+        char mem_total_cmd[]= "cat /proc/meminfo | grep -i memavailable";
+        swap_free_space = meminfo(swap_cmd);
+        total_available_mem = meminfo(mem_total_cmd);
+        total_free_space = swap_free_space + total_available_mem;
+        total_free_mm = total_free_space/(1024*1024);
+        map_chunk_size_gb = MAP_CHUNK_SIZE / 1024 / 1024 / 1024;
+        if(map_chunk_size_gb > total_free_mm)
+        {
+                unsigned long mm_aloc = total_free_mm - 2;
+                unsigned long map_chunk_size = mm_aloc*1024*1024*1024;
```
Comment (sacsant) [inline memory/vatest.py.data/va_test.c:361]: Do we really need these 2 additional variables. We should be able to change this to
MAP_CHUNK_SIZE = (total_free_mm - 2)*1024*1024*1024

File: memory/vatest.py.data/va_test.c [hunk 4]
```diff
@@ -324,10 +324,50 @@ void mixed_position_alterhuge(int chunks1, int chunks2, int chunks3)
...
+                MAP_CHUNK_SIZE = map_chunk_size;
+        }
+}
+
 int main(int argc, char *argv[])
 {
 	int option = 0;
 	int scenario = 0, def_chunks = 0, chunks = 0, chunks_2 = 0;
+	mem_checkpoint();
```
Comment (sacsant) [inline memory/vatest.py.data/va_test.c:370]: This call should be moved after the check for argc < 3

**Review by sacsant — CHANGES_REQUESTED** [review]

File: memory/vatest.py.data/va_test.c [hunk 5]
```diff
@@ -324,6 +324,44 @@ void mixed_position_alterhuge(int chunks1, int chunks2, int chunks3)
...
+        while( token != NULL )
+        {
+                num = atoi(token);
+                if (num == 0 && token != '0');
+                else{
+                        token = NULL;
+			pclose(file);
+                        return num;
+                }
+
```
Comment (sacsant) [inline memory/vatest.py.data/va_test.c:344]: Remove the blank line.

File: memory/vatest.py.data/va_test.c [hunk 6]
```diff
@@ -324,6 +324,44 @@ void mixed_position_alterhuge(int chunks1, int chunks2, int chunks3)
...
+                token = strtok(NULL, " ");
+        }
```
Comment (sacsant) [inline memory/vatest.py.data/va_test.c:346]: Requires another pclose() if the control reaches here.

File: memory/vatest.py.data/va_test.c [hunk 7]
```diff
@@ -324,6 +324,44 @@ void mixed_position_alterhuge(int chunks1, int chunks2, int chunks3)
...
+        int swap_free_space, total_available_mem, total_free_space, total_free_mm, map_chunk_size_gb;
+        char swap_cmd[] = "cat /proc/meminfo | grep -i swapfree";
+        char mem_total_cmd[]= "cat /proc/meminfo | grep -i memavailable";
+        swap_free_space = meminfo(swap_cmd);
+        total_available_mem = meminfo(mem_total_cmd);
+        total_free_space = swap_free_space + total_available_mem;
+        total_free_mm = total_free_space/(1024*1024);
+        map_chunk_size_gb = MAP_CHUNK_SIZE / 1024 / 1024 / 1024;
+        if(map_chunk_size_gb > total_free_mm)
+        {
```
Comment (sacsant) [inline memory/vatest.py.data/va_test.c:360]: No need to have these brackets

File: memory/vatest.py.data/va_test.c [hunk 8]
```diff
@@ -324,6 +324,44 @@ void mixed_position_alterhuge(int chunks1, int chunks2, int chunks3)
...
+                else{
+                        token = NULL;
+			pclose(file);
+                        return num;
+                }
+
+                token = strtok(NULL, " ");
+        }
+}
+
```
Comment (sacsant) [inline memory/vatest.py.data/va_test.c:348]: Can you add a comment on logic used in this code?

**Comment by sacsant** [comment]
> A corresponding pclose() call is missing.

@SamirMulani , on second thought, do we really need another function for parsing /proc/meminfo? Why can't existing function local_read_meminfo() be reused/extended?

**Comment by SamirMulani** [comment]
> > A corresponding pclose() call is missing.
> 
> @SamirMulani , on second thought, do we really need another function for parsing /proc/meminfo? Why can't existing function local_read_meminfo() be reused/extended?

@sacsant Thank you for the guidance. Yes, we can use the existing local_read_meminfo() function. Updated commit reflects all the changes as per the inputs.

**Comment by sacsant** [comment]
> @sacsant Thank you for the guidance. Yes, we can use the existing local_read_meminfo() function. Updated commit reflects all the changes as per the inputs.
I see that you have incorporated the review changes into a different commit. Can you fold the latest changes into a single commit ?

**Review by sacsant — COMMENTED** [review]
Can you update the commit message to match with the changes proposed? You can use the same text mentioned in the initial comment of this PR. 

Rest of the changes LGTM.

**Review by sacsant — APPROVED** [review]
Can you post run logs for reference.

LGTM.

**Comment by SamirMulani** [comment]
```
**Avocado test case run log:**
# avocado run --test-runner runner vatest.py
JOB ID     : 0c4e21e92db037e1b2bc46ae1cdb99f0afebd5b4
JOB LOG    : /root/avocado/job-results/job-2022-06-29T00.35-0c4e21e/job.log
 (1/1) vatest.py:VATest.test: PASS (2.00 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /root/avocado/job-results/job-2022-06-29T00.35-0c4e21e/results.html
JOB TIME   : 2.45 s
```

**Va test binary run log:**
```
 ./va_test -s 1 -n 0 -h 0 -d 0

Scenario 1 : Alloc default pagesize from 0-512 tb VA with above 128 first

Allocating default pagesize pages > 128TB 
> 128Tb: 
 chunks allocated= 24575 
Allocating default pagesize pages < 128TB 
< 128Tb: 
 chunks allocated= 8190 

Test Passed!!
```

**Pull request merged**